ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/PrimitiveConvExtensions.kt

```diff
@@ -0,0 +1,193 @@
+        }
+    }
+    val result = PrimitiveArray(resultShape.shapeSize())
+    val outputShape = IntArray(resultShape.size - 2) { resultShape[it + 2] }
+
+    val kernel = IntArray(w.shape.size - 2) { w.shape[it + 2] }
+    val xOffset = shape[1] / groups * inputShape.shapeSize()
+    val yOffset = resultShape.shapeSize() / resultShape[0] / groups
+    val wOffset = w.shape.shapeSize() / groups
+    val kernelDim = shape[1] / groups * kernel.shapeSize()
+    val col = PrimitiveArray(outputShape.shapeSize() * kernelDim)
+
+    val xPointer = array.pointer()
+    var yPointer = 0
+
+    for (imageId in 0 until shape[0]) {
+        for (groupId in 0 until groups) {
+            val prevLiX = xPointer.linearIndex
+            xPointer.linearIndex += xOffset * groupId
+            if (kernel.size == 2) {
+                primitiveIm2Col(
+                    xPointer,
+                    shape[1] / groups,
+                    inputShape[0],
+                    inputShape[1],
+                    kernel[0],
+                    kernel[1],
+                    dilations[0],
+                    dilations[1],
+                    pads[0],
+                    pads[1],
+                    pads[2],
+                    pads[3],
+                    strides[0],
+                    strides[1],
+                    col
+                )
+            } else {
+                primitiveIm2Col(
+                    xPointer,
+                    inputShape,
+                    outputShape,
+                    kernelDim,
+                    kernel,
+                    strides,
+                    dilations,
+                    pads,
+                    kernel.size,
+                    col
+                )
+            }
+            xPointer.linearIndex = prevLiX
+
+            val prevLiY = yPointer
+            val wPointer = w.array.pointer(wOffset * groupId)
+            yPointer += yOffset * groupId
+            gemmConvCoroutines(
+                wPointer,
+                col,
+                yPointer,
+                result,
+                w.shape[0] / groups,
+                outputShape.shapeSize(),
+                kernelDim
+            )
+            yPointer = prevLiY
+        }
+
+        if (imageId != shape[0] - 1) {
+            xPointer.linearIndex += xOffset * groups
+            yPointer += yOffset * groups
+        }
+    }
+
+    if (b != null) {
+        var resultPointer = 0
+        repeat(shape[0]) {
+            val bPointer = PrimitivePointer(b.array)
+            repeat(w.shape[0]) {
+                repeat(outputShape.shapeSize()) {
+                    result[resultPointer] += bPointer.get()
+                    resultPointer++
+                }
+                bPointer.increment()
+            }
+        }
+    }
+
+    return PrimitiveNDArray(resultShape) { i -> result[i].toPrimitive() }
+}
+
+@SpecifyPrimitives(include = [])
+fun IntArray.shapeSize(): Int {
+    return this.reduce { size, i -> size * i }
```

Comment:
This is super-specific function, so it shouldn't be public but internal at least. Also it will throw an exception if used with an empty shape (scalar tensors have empty shape). Maybe replace it with fold with initial value = 1?

Reply:
And maybe rename it to smth like inferLinearSize for more readability